Rename CMake options and C macros to use PUMI_ prefix

bobpaw opened the pull request:
## Rename CMake options and C macros to use PUMI_ prefix

I prefixed lots of options with PUMI_ and renamed variables/options with the SCOREC_ prefix which are not created by bob.cmake. In general I set the default value for options like `PUMI_ENABLE_SIMMETRIX` to `${ENABLE_SIMMETRIX}` which allows us to remain somewhat backward compatible and also inherit the selection from parent projects by default. 

To find things that needed to be changed, I used the following searches:

```
git grep '\WENABLE' :^Doxyfile.in :^Doxyfile_internal.in
git grep '\WHAS' :^Doxyfile.in :^Doxyfile_internal.in
git grep '\WHAVE' :^Doxyfile.in :^Doxyfile_internal.in
git grep '\WUSE' :^Doxyfile.in :^Doxyfile_internal.in
```

I removed the private compile definition `-DHAVE_SIMADVMESHING` from phasta/CMakeLists.txt and some test exes and replaced it with the `PUMI_HAS_SIMADVMESHING` macro in apf_sim/apf_simConfig.h. There are some options which are left, specifically Simmetrix things like `SIM_PARASOLID`, `SIM_ACIS`, etc that I wasn't sure how we wanted to deal with yet.

~~An exception that I did not remove is ENABLE_VIZ in test/CMakeLists.txt:~~ Removed in #531.

I did not yet change the project name to SCOREC because I wanted to make sure all of these changes work.

Leaving in draft mode for now because there are a few things (Phasta ~~HAVE_CLOCK_/USE_PCU_TIME~~, LION_COMPRESS, etc) left that I did not get to or am not sure if need to be updated.

Closes #337. Closes #477.

bobpaw committed:
cmake: rename PUMI_ENABLE_SIMMETRIX

- CMakeLists.txt: rename ENABLE_SIMMETRIX to PUMI_ENABLE_SIMMETRIX as
  requested in #337.
- allow unscoped ENABLE_SIMMETRIX to also enable it for PUMI.
- CMakeLists.txt, apf_sim/CMakeLists.txt, gmi_sim/CMakeLists.txt,
  phasta/CMakeLists.txt, python_wrappers/CMakeLists.txt,
  test/CMakeLists.txt, test/testing.cmake: replace all the
  ENABLE_SIMMETRIXs with PUMI_ENABLE_SIMMETRIX.

Signed-off-by: Aiden Woodruff <woodra@rpi.edu>

bobpaw committed:
rename HAVE_SIMMETRIX to PUMI_HAS_SIMMETRIX

- CMakeLists.txt: change HAVE_SIMMETRIX compile definition to
  PUMI_HAS_SIMMETRIX.

- git grep HAVE_SIMMETRIX | cut -d: -f1 | uniq | xargs sed -i
  's/HAVE_SIMMETRIX/PUMI_HAS_SIMMETRIX/'

Signed-off-by: Aiden Woodruff <woodra@rpi.edu>

bobpaw committed:
rename SCOREC_NO_MPI TO PUMI_NO_MPI

- git grep SCOREC_NO_MPI | cut -d: -f1 | uniq | xargs sed -i
  's/SCOREC_NO_MPI/PUMI_NO_MPI/'
- doc/myprogram.cpp: remove #ifdef PUMI_NO_MPI by using pcu::Init.

Signed-off-by: Aiden Woodruff <woodra@rpi.edu>

bobpaw committed:
prefix flags for CGNS and Omega_h and rename NOMPI

- CMakeLists.txt: rename ENABLE_CGNS to PUMI_ENABLE_CGNS and use old
  ENABLE_CGNS as the default value.
- rename ENABLE_OMEGA_H to PUMI_ENABLE_OMEGA_H and use old name as
  default value.
- change all conditionals to use the new scoped name.
- rename SCOREC_NO_MPI to PUMI_NO_MPI.
- change PUMI_ENABLE_SIMMETRIX to use ENABLE_SIMMETRIX as the default
  value like with cgns/omega_h.
- don't rename SCOREC_USE_Omega_h_DEFAULT yet because that is input for
  bob.cmake based on the project name.
- apf/CMakeLists.txt: add apfCGNSempty.cc file when CGNS is not enabled.
- apf/pkg_tribits.cmake: same thing as CMakeLists.txt.
- apf/apfCGNS.cc: remove HAVE_CGNS conditionals since the whole file is
  conditionally compiled based on PUMI_ENABLE_CGNS.
- apf/apfCGNSempty.cc: move #else cases for HAVE_CGNS here so that they
  are compiled. otherwise they never would be.
- cmake/bob.cmake: add documentation
- mds/CMakeLists.txt: add mdsCGNSempty.cc without PUMI_ENABLE_CGNS.
- mds/pkg_tribits.cmake: same thing.
- mds/mdsCGNS: remove #ifdef HAVE_CGNS since the whole file is
  conditionally compiled. move #else branches to mdsCGNSempty.cc.
- mds/mdsCGNSempty.cc: new file with empty functions that always fail.
- test/CMakeLists.txt: rename ENABLE_CGNS to PUMI_ENABLE_CGNS and
  ENABLE_OMEGA_H to PUMI_ENABLE_OMEGA_H.
- test/testing.cmake: same thing.

Signed-off-by: Aiden Woodruff <woodra@rpi.edu>

bobpaw committed:
add PUMI_ prefix to ENABLE_FPP

- CMakeLists.txt: rename option to PUMI_ENABLE_FPP and make default
  value ENABLE_FPP for compatibility.
- rename preprocessor definition to PUMI_DO_FPP.
- ma/maSnapper.cc: use PUMI_DO_FPP.

Signed-off-by: Aiden Woodruff <woodra@rpi.edu>

bobpaw committed:
prefix simmetrix feature options/macros

- CMakeLists.txt: add PUMI_ prefix to simmetrix feature macros
  (fieldsim and advanced meshing).
- apf_sim/apfSIM.cc: use prefixed macros.
- pumi_simModSuiteConfig.h.in: use prefixed macros.

Signed-off-by: Aiden Woodruff <woodra@rpi.edu>

bobpaw committed:
add PUMI_ prefix to ENABLE_DSP option

Signed-off-by: Aiden Woodruff <woodra@rpi.edu>

bobpaw committed:
add PUMI_ prefix to ENABLE_METIS

- metis/CMakeLists.txt: update option name and if statements.
- ma/maInput.cc: update error message to reference new option name.
- test/CMakeLists.txt: update if statements.
- test/testing.cmake: update if statements.

Signed-off-by: Aiden Woodruff <woodra@rpi.edu>

bobpaw committed:
add PUMI_ prefix to STK options

- stk/CMakeLists.txt: add PUMI_ prefix to ENABLE_STK and
  ENABLE_STK_MESH.
- stk/apf_stkConfig.h.in: add PUMI_ prefix to HAS_STK and use
  PUMI_ENABLE_STK_MESH.
- stk/apfExodusOutput.cc: change HAS_STK to PUMI_HAS_STK.
- stk/apfMeshSTK.cc: use PUMI_HAS_STK.
- stk/apfSTK.cc: use PUMI_HAS_STK.

Signed-off-by: Aiden Woodruff <woodra@rpi.edu>

bobpaw committed:
add PUMI_ prefix to zoltan options

- zoltan/CMakeLists.txt: add PUMI_ prefix to ENABLE_ZOLTAN and make
  ENABLE_ZOLTAN value the default to inherit and for compatability.
- update message and if statements.
- zoltan/pkg_tribits.cmake: change variable to PUMI_ENABLE_ZOLTAN.
- test/capVol.cc (makeSplitter): fix typo in zoltan call.
- ma/maInput.cc: update error message when trying to run MeshAdapt with
  Zoltan but without compiling the support.

Signed-off-by: Aiden Woodruff <woodra@rpi.edu>

bobpaw committed:
fix ENABLE_ZOLTAN in test/ CMake files

- test/CMakeLists.txt: rename ENABLE_ZOLTAN to PUMI_ENABLE_ZOLTAN
  wherever found.
- test/testing.cmake: replace ENABLE_ZOLTAN to PUMI_ENABLE_ZOLTAN.

Signed-off-by: Aiden Woodruff <woodra@rpi.edu>

bobpaw committed:
rename HAVE_SIMADVMESHING to PUMI_HAS_SIMADVMESHING

- apf_sim/CMakeLists.txt: rename PUMI_USE_SIMADVMESHING to
  PUMI_HAS_SIMADVMESHING. Cameron (in 330ec29) and then I (in d92cc921)
  changed it because I could not find the input variable
  HAVE_SIMADVMESHING. turns out it was the name
  being used as a C macro.
- apf_sim/apf_simConfig.h.in: change #cmakedefine line
- phasta/CMakeLists.txt: remove HAVE_SIMADVMESHING from compile
  definitions since it should now be read from apf_simConfig.h.
- phasta/chef.cc: #include apf_simConfig.h and replace
  HAVE_SIMADVMESHING with PUMI_HAS_SIMADVMESHING.
- phasta/cut_interface.cc: #include apf_simConfig.h and replace
  HAVE_SIMADVMESHING with PUMI_HAS_SIMADVMESHING.
- phasta/phMeshQuality.cc: #include apf_simConfig.h and replace
  HAVE_SIMADVMESHING with PUMI_HAS_SIMADVMESHING.
- test/CMakeLists.txt: remove HAVE_SIMADVMESHING as a private compile
  definition since it shoudl be read from apf_simConfig.h
- test/measureAnisoStats.cc: #include apf_simConfig.h and replace
  HAVE_SIMADVMESHING with PUMI_HAS_SIMADVMESHING.
- test/measureIsoStats.cc: #include apf_simConfig.h and replace
  HAVE_SIMADVMESHING with PUMI_HAS_SIMADVMESHING.
- test/sim_part.cc: #include apf_simConfig.h and replace
  HAVE_SIMADVMESHING with PUMI_HAS_SIMADVMESHING.

Signed-off-by: Aiden Woodruff <woodra@rpi.edu>

bobpaw committed:
add PUMI_ prefix to DEBUG_FPP macro

- ma/maSnapper.cc: Angel added DEBUG_FPP in 62b50f20a. Change it to
  PUMI_DEBUG_FPP.

- CMakeLists.txt: replace PUMI_ENABLE_FPP with PUMI_DEBUG_FPP since
  DO_FPP no longer exists.

Signed-off-by: Aiden Woodruff <woodra@rpi.edu>

bobpaw committed:
prefix phasta macros with PUMI_and update ph cmake

- phasta/CMakeLists.txt: add PUMI_ prefix to cmake variables and macros.
- remove add_definitions and change INTERFACE to PUBLIC.
- the effect of doing add_definitions and then adding INTERFACE
  definitions is the same as adding PRIVATE and then INTERFACE. adding
  both PRIVATE and INTERFACE is the same as PUBLIC.

- phasta/phiotimer.h, phasta/phiotimer.cc: use PUMI_ prefixed macros.

Signed-off-by: Aiden Woodruff <woodra@rpi.edu>

bobpaw committed:
add deprecation and compatibility for pumi_no_mpi

Signed-off-by: Aiden Woodruff <woodra@rpi.edu>

bobpaw commented:
I found what seems like a reasonable solution on the CMake discourse forum:

https://discourse.cmake.org/t/exporting-renamed-multiple-named-configs/2586

For us, the recommendation is to distribute a second `SCORECConfig.cmake` which exports all our targets pointing from `SCOREC::<target>` to `PUMI::<target>` depending on the requested version of PUMI.

jacobmerson commented:
This seems like a reasonable approach. Just came to mind, but do we want to have a specific target version or timespan where support of old naming will be dropped? Even if that doesn't go into the messages, it would be good for us to have an agreed upon/documented idea of when to do this.

bobpaw commented:
The minimum semantic versioning requirement for deprecation is to release a backwards compatible minor version with the deprecation, then a major version with the deprecated elements removed. It's also common to put these changes into a major 5.0 release and later release 6.0 with the old names dropped. I do not know what Cameron would prefer there or with respect to a timeline.

bobpaw committed:
add old non-prefixed simmetrix macros

- pumi_simModSuiteConfig.h.in: replace USE_FIELDSIM and
  USE_SIM_ADVMESHING when the new prefixed ones are defined.

Signed-off-by: Aiden Woodruff <woodra@rpi.edu>

cwsmith commented:
Thank you @bobpaw .  

IIUC, as is (012d184), this PR is backwards compatible.  Is that correct?
If so, do we want to include the project name change (`SCOREC` -> `PUMI`) in this PR, create the backward compatible `SCORECConfig.cmake`, and bump the minor version number to 4.3.0?  As you described above, after some time we can then drop the old project name and create the 5.0 release.  

Note, we already have a 'downstream' proxy app we can test with in the repo here: https://github.com/SCOREC/core/tree/master/doc

bobpaw commented:
> IIUC, as is ([012d184](https://github.com/SCOREC/core/commit/012d184b841ce173ca4ba467544e54019c866227)), this PR is backwards compatible. Is that correct?

That is the intention, and I believe it is correct. Will need to do a final review to confirm.

bobpaw commented:
> If so, do we want to include the project name change (`SCOREC` -> `PUMI`) in this PR, create the backward compatible `SCORECConfig.cmake`, and bump the minor version number to 4.3.0? As you described above, after some time we can then drop the old project name and create the 5.0 release.

Sounds good. I will get on that.

bobpaw committed:
rename project and bump minor version

- CMakeLists.txt: rename project and bump minor version.
- rename bob.cmake type variables to use PUMI as project name.
- add SCORECConfig.cmake and SCORECConfigVersion.cmake after adding
  PUMIConfig.cmake.

- cmake/SCORECConfig.cmake.in: hand-coded transition package
  configuration file.
- should export all PUMI:: targets as SCOREC:: targets and alert users
  to change the project name that they use in find_package.
- https://discourse.cmake.org/t/exporting-renamed-multiple-named-configs/2586/2

Signed-off-by: Aiden Woodruff <woodra@rpi.edu>

bobpaw committed:
be compatible with SCOREC_CXX_OPTIMIZE, etc. from bob

- CMakeLists.txt: add compatibility with and deprecation notices for
  SCOREC_CXX_OPTIMIZE, SCOREC_CXX_SYMBOLS, SCOREC_CXX_WARNINGS,
  SCOREC_CXX_FLAGS, and SCOREC_EXTRA_CXX_FLAGS.

Signed-off-by: Aiden Woodruff <woodra@rpi.edu>
